Foundation/src/ActiveThreadPool.cpp

```diff
@@ -15,347 +15,493 @@
+		skipToNextOrEnd();
+
+		return runnable;
+	}
+
+	int priority() const
+	{
+		return _priority;
 	}
 
 private:
-	std::atomic<Thread::Priority> _priority;
-	Runnable& _runnable;
-	std::string _name;
+	int _priority = 0;
+	int _firstIndex = 0;
+	int _lastIndex = -1;
+	Runnable* _entries[MaxPageSize];
 };
 
-class ActiveThread: public Runnable
+
+class ActivePooledThread: public Runnable
 {
 public:
-	ActiveThread(const std::string& name, int stackSize = POCO_THREAD_STACK_SIZE);
-	~ActiveThread() override = default;
+	ActivePooledThread(ActiveThreadPoolPrivate* pool);
 
 	void start();
-	void start(Thread::Priority priority, Runnable& target);
-	void start(Thread::Priority priority, Runnable& target, const std::string& name);
 	void join();
-	void release();
-	void run() override;
+	bool isRunning() const;
+
+	void setRunnable(Runnable& target);
+	void notifyRunnableReady();
+	void registerThreadInactive();
+
+	virtual void run() override;
 
 private:
-	NotificationQueue _pTargetQueue;
-	std::string       _name;
-	Thread            _thread;
-	Event             _targetCompleted;
-	FastMutex         _mutex;
-	const long        JOIN_TIMEOUT = 10000;
-	std::atomic<bool> _needToStop{false};
+	ActiveThreadPoolPrivate* _pool;
+	Runnable* _pTarget;
+	Condition _runnableReady;
+	Thread _thread;
 };
 
 
-ActiveThread::ActiveThread(const std::string& name, int stackSize):
-	_name(name),
-	_thread(name),
-	_targetCompleted(Event::EVENT_MANUALRESET)
+class ActiveThreadPoolPrivate
 {
-	poco_assert_dbg (stackSize >= 0);
-	_thread.setStackSize(stackSize);
+public:
+	ActiveThreadPoolPrivate(int capacity, int stackSize);
+	ActiveThreadPoolPrivate(int capacity, int stackSize, const std::string& name);
+	~ActiveThreadPoolPrivate();
+
+	bool tryStart(Runnable& target);
+	void enqueueTask(Runnable& target, int priority = 0);
+	void startThread(Runnable& target);
+	void joinAll();
+
+	int activeThreadCount() const;
+
+public:
+	mutable FastMutex mutex;
+	std::string name;
+	std::set<ActivePooledThread*> allThreads;
+	std::list<ActivePooledThread*> waitingThreads;
+	std::list<ActivePooledThread*> expiredThreads;
+	std::list<QueuePage*> runnables;
+	Condition noActiveThreads;
+
+	int expiryTimeout = 30000;
+	int maxThreadCount;
+	int stackSize;
+	int activeThreads = 0;
+	int serial = 0;
+};
+
+
+ActivePooledThread::ActivePooledThread(ActiveThreadPoolPrivate* pool):
+	_pool(pool),
+	_pTarget(nullptr)
+{
+	poco_assert(_pool != nullptr);
+	std::ostringstream name;
+	name << _pool->name << "[#" << ++_pool->serial << "]";
+	_thread.setName(name.str());
+	_thread.setStackSize(_pool->stackSize);
 }
 
-void ActiveThread::start()
+
+void ActivePooledThread::start()
 {
-	_needToStop = false;
 	_thread.start(*this);
 }
 
 
-void ActiveThread::start(Thread::Priority priority, Runnable& target)
+void ActivePooledThread::setRunnable(Runnable& target)
 {
-	_pTargetQueue.enqueueNotification(Poco::makeAuto<NewActionNotification>(priority, target, _name));
+	_pTarget = &target;
```

Comment:
@siren186 , can you take a look at this warning?

Reply:
That is because I tried to assign `Runnable& target` to `Runnable* _pTarget`.
I used the same code as `Poco::ThreadPool`:
https://github.com/pocoproject/poco/blob/e4aaeacea8872e9361f3d67217c5a254cf6fc773/Foundation/src/ThreadPool.cpp#L89

I do not known how to fix this warning.

Reply:
This warning can be fixed by `std::optional<std::reference_wrapper<Runnable>> _pTarget`.
But unfortunately it cannot be compiled by macOS.
see: https://github.com/pocoproject/poco/issues/4650